### PR TITLE
Feature/gh 552 logs store

### PR DESCRIPTION
**Pull request by stebenoist**
# Pull Request description
Refactor logs package to allow to use new implementations.

### What I did

This PR contains https://github.com/ystia/yorc/pull/580.
And:

- Events package (only stuff about logs) refactoring
- Encryptor utility for encrypt/decrypt data
- New Store interface methods to handle logs:
  - GetLastModifyIndex
  - List

- Storage configuration management
It allows to configure stores of log, event and deployment type.
- Update documentation

### How to verify it
- Read the doc !
- Add in Yorc config file the lines:

```
  "storage": {
    "stores": [
      {
        "name": "myCipherFileStore",
        "implementation": "cipherFileCache",
        "types":  ["Log"],
        "properties": {
          "secret_key": "myverystrongpasswordo32bitlength"
        }
      }
    ]
  }
```

- Check correct config is saved in Consul
- Deploy an app with the related yorc a4c plugin https://github.com/alien4cloud/alien4cloud-yorc-provider/pull/15)
- Check logs are encrypted and stored in work/store directory.
- Check logs are correctly polled from A4C and Yorc CLI

### Description for the changelog
Refactor logs to allow to plug new implementations (GH-552](https://github.com/ystia/yorc/issues/552))
## Applicable Issues
#552 
#554

### CODE REVIEW

File: .codeclimate.yml [hunk 1]
```diff
@@ -1,4 +1,6 @@
 version: "2"         # required to adjust maintainability checks
+exclude_patterns:
+  - "**/*_test.go"
```
Comment (loicalbertin) [inline .codeclimate.yml]: Why? :smile:

File: events/log_entries.go [hunk 2]
```diff
@@ -29,6 +31,7 @@ import (
 
 // LogEntry is the log entry representation
 type LogEntry struct {
+	ctx            context.Context
```
Comment (loicalbertin) [inline events/log_entries.go]: Why storing context here?
Except for a very specific good reason this is a code smell: 
From https://github.com/golang/go/wiki/CodeReviewComments#contexts

> Don't add a Context member to a struct type; instead add a ctx parameter to each method on that type that needs to pass it along. The one exception is for methods whose signature must match an interface in the standard library or in a third party library.

And from https://golang.org/pkg/context/ :
> Do not store Contexts inside a struct type; instead, pass a Context explicitly to each function that needs it. The Context should be the first parameter, typically named ctx:

File: storage/internal/file/store.go [hunk 3]
```diff
@@ -0,0 +1,462 @@
...
+		withCache:         withCache,
+		withEncryption:    withEncryption,
+	}
+
+	// Instantiate cache if necessary
+	if withCache {
+		fs.cache, err = ristretto.NewCache(&ristretto.Config{
+			NumCounters: 1e7,     // number of keys to track frequency of (10M).
+			MaxCost:     1 << 30, // maximum cost of cache (1GB).
+			BufferItems: 64,      // number of keys per Get buffer.
```
Comment (loicalbertin) [inline storage/internal/file/store.go]: All of this should be parametric.

File: storage/internal/file/store.go [hunk 4]
```diff
@@ -0,0 +1,462 @@
...
+		fp := s.buildFilePath(k, true)
+		fInfo, err = os.Stat(fp)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return 0, errors.Wrapf(err, "failed to get last index for key:%q", k)
+			}
+			// File not found : the key doesn't exist
+			return 0, nil
+		}
+		return uint64(fInfo.ModTime().UnixNano()), nil
```
Comment (loicalbertin) [inline storage/internal/file/store.go]: Yes but what if I `touch` the file?

File: storage/internal/file/store.go [hunk 5]
```diff
@@ -0,0 +1,462 @@
...
+}
+
+func (s *fileStore) List(ctx context.Context, k string, waitIndex uint64, timeout time.Duration) ([]store.KeyValueOut, uint64, error) {
+	if waitIndex == 0 {
+		return s.list(k)
+	}
+
+	// Default timeout to 5 minutes
+	if timeout == 0 {
+		timeout = 5 * time.Minute
```
Comment (loicalbertin) [inline storage/internal/file/store.go]: Not configurable?

File: storage/store/test.go [hunk 6]
```diff
@@ -46,10 +62,49 @@ func handleGetError(t *testing.T, err error, found bool) {
 	}
 }
 
+// NewTestConsulInstance allows to provide new Consul instance for tests
+// This is a private Consul server instantiation as done in github.com/ystia/yorc/v4/testutil
+// This allows avoiding cyclic dependencies with deployments store package
+func NewTestConsulInstance(t testing.TB) (*testutil.TestServer, *api.Client, config.Configuration) {
```
Comment (loicalbertin) [inline storage/store/test.go]: Same as above looks like messy to define config within this function

File: storage/store/test.go [hunk 7]
```diff
@@ -46,10 +62,49 @@ func handleGetError(t *testing.T, err error, found bool) {
...
+		c.Args = []string{"-ui"}
+		c.LogLevel = logLevel
+	}
+
+	srv1, err := testutil.NewTestServerConfig(cb)
+	if err != nil {
+		t.Fatalf("Failed to create consul server: %v", err)
+	}
+
+	workingDir, err := ioutil.TempDir("/tmp", "work")
```
Comment (loicalbertin) [inline storage/store/test.go]: Same as above consider using `./testdata/` instead of `/tmp`

File: storage/store_mgr.go [hunk 8]
```diff
@@ -15,79 +15,246 @@
...
+		stores[storeType] = consul.NewStore()
+	default:
+		log.Printf("[WARNING] unknown store implementation:%q. This will be ignored.", storeType)
+	}
 	return nil
 }
 
 // GetStore returns the store related to a defined store type
-func GetStore(typ types.StoreType) Store {
+func GetStore(typ types.StoreType) store.Store {
```
Comment (loicalbertin) [inline storage/store_mgr.go]: Maybe use check existence form (the one that also returns a bool) of the get key in map to validate WDYT?

File: testutil/helper.go [hunk 9]
```diff
@@ -57,17 +60,21 @@ func NewTestConsulInstanceWithConfigAndStore(t testing.TB, cb testutil.ServerCon
 //  - starts a Consul Publisher
 //  - stores common-types to Consul only if storeCommons bool parameter is true
 // Warning: You need to defer the server stop command in the caller
-func NewTestConsulInstanceWithConfig(t testing.TB, cb testutil.ServerConfigCallback, storeCommons bool) (*testutil.TestServer, *api.Client) {
+func NewTestConsulInstanceWithConfig(t testing.TB, cb testutil.ServerConfigCallback, storeCommons bool) (*testutil.TestServer, *api.Client, string) {
```
Comment (loicalbertin) [inline testutil/helper.go]: Same as above

File: testutil/helper.go [hunk 10]
```diff
@@ -57,17 +60,21 @@ func NewTestConsulInstanceWithConfigAndStore(t testing.TB, cb testutil.ServerCon
...
 	srv1, err := testutil.NewTestServerConfig(cb)
 	if err != nil {
 		t.Fatalf("Failed to create consul server: %v", err)
 	}
 
+	workingDir, err := ioutil.TempDir("/tmp", "work")
```
Comment (loicalbertin) [inline testutil/helper.go]: same as above

File: .codeclimate.yml (hunk 1 above)
Comment (stebenoist) [reply .codeclimate.yml]: Too many "similar code", "complexity" in tests code, we should exclude them from CodeClimate. And I need to change pattern for storage/store/test.go !

File: CHANGELOG.md [hunk 11]
```diff
@@ -27,6 +27,7 @@ func GetWorkflow(ctx context.Context, deploymentID, workflowName string) (*tosca
 
 * Should be able to bootstrap Yorc on OpenStack with Identity API v3 ([GH-575](https://github.com/ystia/yorc/issues/575))
 * Refactor deployments package to be able to use different storage backends - part Two: Consul as default Deployments store implementation ([GH-530](https://github.com/ystia/yorc/issues/530))
+* Refactor logs to allow to plug new implementations (GH-552](https://github.com/ystia/yorc/issues/552))
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply CHANGELOG.md]: No, it's no more relevant...

File: deployments/store/definition_store_test.go [hunk 12]
```diff
@@ -48,7 +52,7 @@ func TestRunDefinitionStoreTests(t *testing.T) {
 // newTestConsulInstance creates and configures Consul instance
 // for testing functions in the store package
 // Remarque: can't use util functions from testutil package in order to avoid import cycles
-func newTestConsulInstance(t *testing.T) (*testutil.TestServer, *api.Client) {
+func newTestConsulInstance(t *testing.T) (*testutil.TestServer, *api.Client, string) {
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply deployments/store/definition_store_test.go]: Yes but we need to clean up working dir after test ?

File: doc/configuration.rst [hunk 13]
```diff
@@ -1109,6 +1109,95 @@ In this case, Yorc gives priority to the application provided properties.
...
+- ``Log``: this represents the applicative logs present in Alien4Cloud logs. ``consul`` is the default implementation for this store type.
+
+If you face some Consul memory usage issues, you can choose ``fileCache`` or ``cipherFileCache`` as logs may contains private information.
+
+- ``Event``: this represents the applicative events present in Alien4Cloud events. ``consul`` is the default implementation for this store type.
+
+
+Here is a JSON example of stores configuration with a cipherFileStore implementation for logs.
+
+.. code-block:: JSON
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply doc/configuration.rst]: Yep, I put both.

File: storage/store_mgr.go (hunk 8 above)
Comment (stebenoist) [reply storage/store_mgr.go]: Makes sense

File: storage/store/test.go (hunk 7 above)
Comment (stebenoist) [reply storage/store/test.go]: ok

File: storage/store_mgr.go (hunk 8 above)
Comment (stebenoist) [reply storage/store_mgr.go]: But at this point, as each store is mandatory, we should panic if the store is missing

Comment (stebenoist) [reply storage/store_mgr.go]: Or just return the bool if later we handle optional stores ?

Comment (stebenoist) [reply storage/store_mgr.go]: But it makes the code more complex for the caller:
``

File: deployments/store/definition_store_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (loicalbertin) [reply deployments/store/definition_store_test.go]: Don't we need a separate helper function to setup testing stores? This is not really/always related to consul.

File: storage/store_mgr.go (hunk 8 above)
Comment (loicalbertin) [reply storage/store_mgr.go]: https://github.com/golang/go/wiki/CodeReviewComments#dont-panic

File: doc/configuration.rst [hunk 14]
```diff
@@ -1109,6 +1109,95 @@ In this case, Yorc gives priority to the application provided properties.
...
 See `Working with jobs <https://yorc-a4c-plugin.readthedocs.io/en/latest/jobs.html>`_ for more information.
 
+Storage configuration
+---------------------
+
++----------------------------------+------------------------------------------------------------------+-----------+------------------+-----------------+
+|     Property Name                |                          Description                             | Data Type |   Required       | Default         |
+|                                  |                                                                  |           |                  |                 |
++==================================+==================================================================+===========+==================+=================+
+| ``reset``                        | If set to True, reset the stores configuration                   | boolean   | no               |   False         |
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply doc/configuration.rst]: It's explained below.

``
Stores configuration is saved once when Yorc server starts. If you want to re-initialize storage, you have to set the ``reset`` property to True and restart Yorc.

File: doc/configuration.rst [hunk 15]
```diff
@@ -1109,6 +1109,95 @@ In this case, Yorc gives priority to the application provided properties.
...
+        "properties": {
+          "root_dir": "/mypath/to/store",
+          "passphrase": "myverystrongpasswordo32bitlength"
+        }
+      }
+    ]
+  }
+    ....
+
+Stores configuration is saved once when Yorc server starts. If you want to re-initialize storage, you have to set the ``reset`` property to True and restart Yorc.
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply doc/configuration.rst]: Yep, I add the description in the tab

File: events/log_entries.go (hunk 2 above)
Comment (stebenoist) [reply events/log_entries.go]: Hey, this is a very specific good reason as the log entry is built with an existing context. No sense to have another context parameter in the register method

File: storage/internal/file/store.go (hunk 4 above)
Comment (stebenoist) [reply storage/internal/file/store.go]: The modifyIndex is updated as if you modify a key in Consul UI

File: storage/store_mgr.go (hunk 8 above)
Comment (stebenoist) [reply storage/store_mgr.go]: This is not a "normal" error as the stores have to exist. Panic is relevant in this case.

In the future, if we handle optional store, we may have to change this but not sure to need it ?

File: doc/configuration.rst (hunk 15 above)
[the remark this answers is not in the document]
Comment (stebenoist) [reply doc/configuration.rst]: OK

File: deployments/store/definition_store_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (stebenoist) [reply deployments/store/definition_store_test.go]: Right. We should just pass the config to NewTestConsulInstance

File: storage/store/test.go (hunk 6 above)
Comment (stebenoist) [reply storage/store/test.go]: Fixed

File: storage/store/test.go (hunk 7 above)
Comment (stebenoist) [reply storage/store/test.go]: Fixed

File: testutil/helper.go (hunk 9 above)
Comment (stebenoist) [reply testutil/helper.go]: Fixed

File: testutil/helper.go (hunk 10 above)
Comment (stebenoist) [reply testutil/helper.go]: Fixed

File: deployments/store/definition_store_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (stebenoist) [reply deployments/store/definition_store_test.go]: Fixed

File: deployments/store/definition_store_test.go [hunk 16]
```diff
@@ -62,11 +66,15 @@ func newTestConsulInstance(t *testing.T) (*testutil.TestServer, *api.Client) {
 		t.Fatalf("Failed to create consul server: %v", err)
 	}
 
+	workingDir, err := ioutil.TempDir("/tmp", "work")
```
[the remark this answers is not in the document]
Comment (stebenoist) [reply deployments/store/definition_store_test.go]: Hum, with ./testdata, it slows down my IDE as it indexes new files. /tmp was better. Maybe you can check TMPDIR env var for other systems ?

File: deployments/store/definition_store_test.go (hunk 12 above)
[the remark this answers is not in the document]
Comment (loicalbertin) [reply deployments/store/definition_store_test.go]: :+1:

File: deployments/store/definition_store_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (loicalbertin) [reply deployments/store/definition_store_test.go]: Ok maybe use https://golang.org/pkg/os/#TempDir

File: testutil/helper.go [hunk 17]
```diff
@@ -92,3 +91,20 @@ func NewTestConsulInstanceWithConfig(t testing.TB, cb testutil.ServerConfigCallb
 func BuildDeploymentID(t testing.TB) string {
 	return strings.Replace(t.Name(), "/", "_", -1)
 }
+
+// SetupTestConfig sets working directory configuration
+// Warning: You need to defer the working directory removal
+func SetupTestConfig(t testing.TB) config.Configuration {
+	rootDir := "/tmp"
+	if _, err := os.Stat(rootDir); os.IsNotExist(err) {
```
Comment (loicalbertin) [inline testutil/helper.go]: ?????

File: events/log_entries.go (hunk 2 above)
Comment (loicalbertin) [reply events/log_entries.go]: :+1:

File: storage/store_mgr.go (hunk 8 above)
Comment (loicalbertin) [reply storage/store_mgr.go]: :+1:

File: deployments/store/definition_store_test.go (hunk 16 above)
[the remark this answers is not in the document]
Comment (stebenoist) [reply deployments/store/definition_store_test.go]: Done

File: testutil/helper.go (hunk 17 above)
Comment (stebenoist) [reply testutil/helper.go]: Fixed

**Comment by stebenoist** [comment]
Closed as embedded into https://github.com/ystia/yorc/pull/583

**Pull request closed without merging**